iree/compiler/Conversion/LinalgToSPIRV/FoldGPUProcessorIDUses.cpp

```diff
@@ -0,0 +1,243 @@
+    case mlir::AffineExprKind::Add:
+    case mlir::AffineExprKind::Mul: {
+      auto binExpr = expr.cast<AffineBinaryOpExpr>();
+      return doesNotContainDivMod(binExpr.getLHS()) &&
+             doesNotContainDivMod(binExpr.getRHS());
+    };
+    case mlir::AffineExprKind::Mod:
+    case mlir::AffineExprKind::FloorDiv:
+    case mlir::AffineExprKind::CeilDiv:
+      return false;
+    case mlir::AffineExprKind::Constant:
+    case mlir::AffineExprKind::DimId:
+    case mlir::AffineExprKind::SymbolId:
+      return true;
+  }
+}
+
+/// Replaces the given `dim` in `expr` with a constant `value`.
+AffineExpr replaceSymbolWithValue(AffineExpr expr, AffineSymbolExpr dim,
+                                  int64_t value) {
+  auto cstExpr = getAffineConstantExpr(value, expr.getContext());
+  return expr.replace(dim, cstExpr);
+}
+
+/// Converts a dimension string to its corresponding index.
+int dimensionToIndex(StringRef dimension) {
+  return StringSwitch<int>(dimension).Case("x", 0).Case("y", 1).Case("z", 2);
+}
+
+/// Gets the block processor ID's upper bound. This queries the workgroup count
+/// function.
+Optional<int64_t> getProcessorIDUpperBound(gpu::BlockIdOp blockIDOp) {
+  auto numWorkgroupsFn = getNumWorkgroupsFn(blockIDOp.getParentOfType<FuncOp>(),
+                                            getNumWorkgroupsFnAttrName());
+  if (!numWorkgroupsFn) return llvm::None;
+
+  Operation *terminator = numWorkgroupsFn.getBlocks().back().getTerminator();
+  auto retOp = dyn_cast<ReturnOp>(terminator);
+  if (!retOp || retOp.getNumOperands() != 3) return llvm::None;
+  LLVM_DEBUG(llvm::dbgs() << "workgroup count function return op: " << retOp
+                          << "\n");
+
+  int index = dimensionToIndex(blockIDOp.dimension());
+  IntegerAttr attr;
+  if (!matchPattern(retOp.getOperand(index), m_Constant(&attr)))
+    return llvm::None;
+
+  return attr.getInt();
+}
+
+/// Gets the thread processor ID's upper bound. This queries the SPIR-V entry
+/// point ABI.
+Optional<int64_t> getProcessorIDUpperBound(gpu::ThreadIdOp threadIDOp) {
+  FuncOp funcOp = threadIDOp.getParentOfType<FuncOp>();
+  auto abiAttr = funcOp.getAttrOfType<spirv::EntryPointABIAttr>(
+      spirv::getEntryPointABIAttrName());
+  if (!abiAttr) return llvm::None;
+
+  int index = dimensionToIndex(threadIDOp.dimension());
+  auto valueIt = abiAttr.local_size().getIntValues().begin() + index;
+  return (*valueIt).getZExtValue();
+}
+
+/// Folds `affine.min` ops which has only one symbol operand, which is a
+/// processor ID. For such cases we can use the processor ID's upper bound to
+/// simplify the `affine.min`.
+///
+/// For example, this pattern can simplify the following IR:
+/// ```
+/// %id = "gpu.thread_id"() {dimension = "x"} : () -> index
+/// ...
+/// affine.min affine_map<()[s0] -> (3, s0 * -2 + 225)>()[%id]
+/// ```
+/// if the upper bound for thread ID along the x dimension is 112.
+struct FoldAffineMinOverProcessorID : OpRewritePattern<AffineMinOp> {
```

Comment:
Couple of issues I see with this pattern and adding it as a general pass.

In isolation replacing the block ID with its maximum value is wrong cause you will actually change the value of the `affine.min`. So this doesnt seem to be a legal transformation.

Further, you are substituting the upper bound for the thread ID. In the example above if the affine map were

```
affine_map<()[s0] -> (3, 225 - s0)>()[%id]
```

would that approach still hold.

IIUC the reason why we need this "canonicalization" is that the value of this affine map is used to compute the size of the vector needed (it also comes up during promotion cause it determines the size of the workgroup memory buffer). So you are trying to get a constant "over-estimate" of the size. I think we need to account for the uses of the `affine.min` operation, and also think a more general solution needs a more involved analysis of the AffineExpr in the result. 

Hopefully I am not totally on the wrong track here. Please let me know.

Reply:
Good catch! There are indeed issues regarding linear expression check and how we check whether a value should be the min.

If an expression involving the processor ID is linear, then the expression is in the range we get by substituting the processor ID with its min and max. However, I should only check _existing_ constant result in the original map. If that constant is <= the range gotten from processor ID and other constants in the original map, then the constant should be min.

Fixed check around linear expression and min candidates.

Reply:
Yup, exactly. This only handles perfectly divisible cases, which are what I'm trying to handle at the moment. :)

> So the solution is that when we want to estimate size of something (during vectorization or promotion) we handle the affine.min correctly.

Not sure I follow exactly here. Could you explain a bit more? These `affine.min` ops are feeding into subview ops. Without being static, we cannot vectorize. I guess we can explicitly throw in patterns that takes in, e.g., `linalg.conv` on memrefs with dynamic dimensions, and first "over-estimate" by treating the dynamic dimensions caused by `affine.min` as always of size `TS`, and then, just at the last load4/store4 operation perform proper masking. (We still want to load4/store4 because it's fast. But it does mean we need to shift "left" a bit to read some previously-already-read elements and mask them alway, to avoid being out-of-bound and incorrect computation.) That's something I'd like to explore. Is that what you meant?

But the above can still work with the current folding pattern. If this pattern kicks in we don't need all the complicated analysis in the above. So we can have both. :)

Reply:
> > So the solution is that when we want to estimate size of something (during vectorization or promotion) we handle the affine.min correctly.
> 
> Not sure I follow exactly here. Could you explain a bit more? These `affine.min` ops are feeding into subview ops. Without being static, we cannot vectorize. I guess we can explicitly throw in patterns that takes in, e.g., `linalg.conv` on memrefs with dynamic dimensions, and first "over-estimate" by treating the dynamic dimensions caused by `affine.min` as always of size `TS`, and then, just for the boundary tile perform scalar load/store and compose vectors by ourselves. That's something I'd like to explore. Is that what you meant?

- The `affine.min` ops are feeding into the size operands of the subview. These subviews are either operands of the linalg operation being vectorized, or are the subviews that are promoted. So it is better to keep them static. I think what we need is to use size `TS` for the boundary tiles also. You might overcompute/use more workgroup memory.  So we dont need more patterns I think. We just want to make sure we get the size `TS` for ALL tiles.

Reply:
Awesome, I think we are aligned over the general direction here; can figure details out when needed. :)